File: rust/src/managed.rs
use reqwest::Client;
use serde::Deserialize;
use std::env;
use std::sync::Arc;

use crate::{
    errors::L402Error,
    types::{BoxFuture, Invoice, LightningProvider},
};

const SHINYDAPPS_API: &str = "https://l402kit.vercel.app";

pub struct ManagedProvider {
    owner_address: String,
    client: Client,
}

impl ManagedProvider {
    pub fn new(owner_address: String) -> Arc<Self> {
        Arc::new(Self {
            owner_address,
            client: Client::new(),
        })
    }

    async fn do_create_invoice(&self, amount_sats: u64) -> Result<Invoice, L402Error> {
        #[derive(Deserialize)]
        struct InvoiceResponse {
            #[serde(rename = "paymentRequest")]
            payment_request: String,
            #[serde(rename = "paymentHash")]
            payment_hash: String,
            macaroon: String,
        }

        let res = self
            .client
            .post(format!("{SHINYDAPPS_API}/api/invoice"))
            .json(&serde_json::json!({ "amountSats": amount_sats }))
            .send()
            .await
            .map_err(|e| L402Error::InvoiceCreationFailed(e.to_string()))?;

        if !res.status().is_success() {
            return Err(L402Error::InvoiceCreationFailed(format!(
                "API returned {}",
                res.status()
            )));
        }

        let data: InvoiceResponse = res
            .json()
            .await
            .map_err(|e| L402Error::InvoiceCreationFailed(e.to_string()))?;

        Ok(Invoice {
            payment_request: data.payment_request,
            payment_hash: data.payment_hash,
            macaroon: data.macaroon,
            amount_sats,
        })
    }

    /// Fire-and-forget: send split to ShinyDapps backend.
    pub async fn send_split(&self, amount_sats: u64) {
        let _ = self
            .client
            .post(format!("{SHINYDAPPS_API}/api/split"))
            .header("x-split-secret", env::var("SPLIT_SECRET").unwrap_or_default())
            .json(&serde_json::json!({
                "amountSats": amount_sats,
                "ownerAddress": self.owner_address,
            }))
            .send()
            .await;
    }
}

impl LightningProvider for ManagedProvider {
    fn create_invoice<'a>(&'a self, amount_sats: u64) -> BoxFuture<'a, Result<Invoice, L402Error>> {
        Box::pin(async move { self.do_create_invoice(amount_sats).await })
    }
}

File: src/managed.ts
import type { LightningProvider, Invoice } from "./types";

const SHINYDAPPS_API = process.env.SHINYDAPPS_API_URL ?? "https://l402kit.com";
const SPLIT_SECRET   = process.env.SPLIT_SECRET ?? "";

export interface DirectoryRegistration {
  /** Public URL of the L402-protected endpoint (e.g. "https://api.example.com/v1/weather") */
  url: string;
  /** Display name shown in the API directory */
  name: string;
  /** Price in satoshis — must match what the middleware actually charges */
  priceSats: number;
  description?: string;
  category?: "data" | "ai" | "finance" | "weather" | "compute" | "storage" | "other";
}

/**
 * Cloud-managed Lightning provider (l402kit.com hosted service).
 *
 * Invoices and splits are handled server-side. You receive 99.7% of each
 * payment directly to your Lightning Address. Explicit opt-in.
 *
 * For sovereign mode (0% fee), use AlbyProvider, BTCPayProvider, or BlinkProvider.
 *
 * @example
 * ```ts
 * import { ManagedProvider } from "l402-kit";
 * l402({ priceSats: 10, lightning: ManagedProvider.fromAddress("you@blink.sv") })
 * // with directory registration:
 * l402({ priceSats: 10, lightning: ManagedProvider.fromAddress("you@blink.sv", {
 *   registerDirectory: { url: "https://api.you.com/weather", name: "Weather API", priceSats: 10 }
 * }) })
 * ```
 */
export class ManagedProvider implements LightningProvider {
  private constructor(private ownerAddress: string) {}

  static fromAddress(
    address: string,
    opts?: { registerDirectory?: DirectoryRegistration },
  ): ManagedProvider {
    const provider = new ManagedProvider(address);
    if (opts?.registerDirectory) {
      const { url, name, priceSats, description, category } = opts.registerDirectory;
      fetch(`${SHINYDAPPS_API}/api/register`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          url,
          name,
          price_sats: priceSats,
          lightning_address: address,
          description: description ?? undefined,
          category: category ?? "other",
        }),
      }).catch(() => { /* silent — directory registration is best-effort */ });
    }
    return provider;
  }

  async createInvoice(amountSats: number): Promise<Invoice> {
    const res = await fetch(`${SHINYDAPPS_API}/api/invoice`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ amountSats, ownerAddress: this.ownerAddress }),
    });
    if (!res.ok) throw new Error("ManagedProvider: invoice creation failed");
    const data = await res.json() as { paymentRequest: string; paymentHash: string; macaroon: string };
    return { ...data, amountSats, expiresAt: Date.now() + 3_600_000 };
  }

  // ManagedProvider uses Blink webhook for payment confirmation, not polling.
  // Verification happens via SHA256(preimage) == paymentHash in the middleware.
  async checkPayment(): Promise<boolean> { return false; }

  async sendSplit(amountSats: number): Promise<void> {
    const res = await fetch(`${SHINYDAPPS_API}/api/split`, {
      method: "POST",
      headers: { "Content-Type": "application/json", "x-split-secret": SPLIT_SECRET },
      body: JSON.stringify({ amountSats, ownerAddress: this.ownerAddress }),
    });
    if (!res.ok) {
      const body = await res.text().catch(() => "");
      throw new Error(`ManagedProvider split ${res.status}: ${body.slice(0, 120)}`);
    }
  }
}
